### DISCUSSION TITLE
Flat Config feedback from TypeScript-ESLint support

### DISCUSSION THREAD

**Discussion by bradzacher** [eslint/eslint#18635] [General]
Following on from https://x.com/bradzacher/status/1806688883217997897
Just going to collate a set of "bugs" and support questions people have submitted to us over at @typescript-eslint to provide visibility into how users across the ecosystem are (mis)using flat configs and running into problems.

I'm tagging issues where the user had a bad flat config in our repo [with `fix: flat config user error`.](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22fix%3A+flat+config+user+error%22+sort%3Acreated-desc).

Here is a list of the issue and a 1-line summary of the problem:

- https://github.com/typescript-eslint/typescript-eslint/issues/9439
  - User misunderstood how `files` worked and assumed that `files` meant only that one config object would apply to those files.
- https://github.com/typescript-eslint/typescript-eslint/issues/8986
  - User misunderstood the global ignore syntax
- https://github.com/typescript-eslint/typescript-eslint/issues/8787
  - User misunderstood local `ignores` vs global `ignores`
- https://github.com/typescript-eslint/typescript-eslint/issues/8798
  - User misunderstood local `ignores` vs global `ignores`
- https://github.com/typescript-eslint/typescript-eslint/issues/8563
  - User misunderstood local `ignores` vs global `ignores`

And discord support questions:

- https://discord.com/channels/1026804805894672454/1248027293561917512
  - User misunderstood the minimatch syntax and differences in the syntax between the global `ignores `and local `ignores`.
- https://discord.com/channels/1026804805894672454/1250753880300453960
  - User misunderstood local `ignores` vs global `ignores`
- https://discord.com/channels/1026804805894672454/1253126451499237396
  - User misunderstood minimatch syntax
- https://discord.com/channels/1026804805894672454/1254962562605383721
  - User misunderstood how `files` worked and assumed that `files` meant the entire config array would only match those files
- https://discord.com/channels/1026804805894672454/1230628675343614012
  - User misunderstood local `ignores` vs global `ignores`
- https://discord.com/channels/1026804805894672454/1228383886246154241
  - User misunderstood minimatch syntax
- https://discord.com/channels/1026804805894672454/1257597908061126707
  - User did not know how to use `files`/`ignores` with 3rd party configs

----

All of these so far are really just RTFM problems and problems users could have likely eventually solved by using `--inspect-config`.

But there are some common themes we can see amongst them (see separate threads below).

**Comment by bradzacher** [comment]
The big one is the distinction between "global `ignores`" and "local `ignores`".
To me it seems that people find it unintuitive to have special case handling for global ignores -- i.e. a config object with just `ignores` and nothing else. Additionally when used in such a position there is additional syntax (folder matching) available that otherwise wouldn't work.

Perhaps this is something that we might consider changing to clarify things. Two quick spitball ideas:
- introducing a new property (like `globalIgnores`) to clearly differentiate it from local `ignores`
    - This would clear up the confusion between syntax and allow it to be specified on any object.
- introducing a property on the config array itself like `globalIgnores` which avoids the need to pass a special one-property object in the array.

**Reply by Jakobud** [reply]
> there is additional syntax (folder matching) available that otherwise wouldn't work

Why doesn't folder pattern matching work in local ignores? This is super unintuitive. Simple glob-based pattern matching has been around for many years and hasn't had these types of limitations. Did eslint try to create some special custom pattern matching system instead of just using what already has worked?

**Reply by bradzacher** [reply]
I don't know the full context and can't find it written in the RFC - but I believe that this was an optimisation they implemented.

Folder syntax in global ignores were supported because it allows eslint to short-circuit traversal of those folders entirely and avoid doing work.
However a local config object ignore does not change traversal at all and only applicability of an object.
Technically one could support folder syntax in local ignores but it would just be short-hand for `folder/**/*`.

**Reply by nzakas** [reply]
IIRC there was a technical issue wrt to directory matches against files not behaving the way we thought it should. This is definitely a rough edge in flat config that we may revisit. 

Regarding global ignores vs. local ignores -- adding another property would only make things more confusing. What would it mean if a config object has both `ignores` and `globalIgnores`? Surely that would be even harder to make sense of.

I believe the current behavior is actually the most intuitive when considered the context of flag config as a whole. For instance, the following two configs are equivalent:

```js
// applies to all files except foo.js
{
    files: ["**/*"],
    ignores: ["**/foo.js"],
    rules: {
        "no-console": "error"
    }
}

// applies to all files except foo.js
{
    ignores: ["**/foo.js"],
    rules: {
        "no-console": "error"
    }
}
```

So the way this is implemented is that a config object with `ignores` and at least one other property is the same as having `files: ["**/*"]` in that object. If we apply that to config objects without other properties, we get this:

```js
// search for all files but exclude foo.js
{
    files: ["**/*"],
    ignores: ["**/foo.js"]
}
```

This is telling ESLint to match all files except `foo.js`...and then do what with them? Nothing! So for every file that ESLint processes, we should ignore any that are named `foo.js`, effectively distributing the `ignores` to every config object in the array. And if every config object in the array is ignoring this pattern, there's no reason to even read the file.

That's how you end up with global ignores.

**Comment by bradzacher** [comment]
Minimatch syntax is a common sticking point given the change in how it works between legacy configs and flat configs. I know I personally was stung by this change (https://github.com/eslint/eslint/issues/17822)

Some of this could be handled by `--inspect-config` - because it provides a way to test file-paths to see what objects match.
However a lot of people don't seem to know about this tool. I wonder if there's a way we can better promote the tool from the CLI and from the docs themselves to increase visibility and awareness of it?

**Reply by nzakas** [reply]
> I wonder if there's a way we can better promote the tool from the CLI and from the docs themselves to increase visibility and awareness of it?

I think you're correct, and a "Debugging Configs" page is probably called for.

**Comment by nzakas** [comment]
Thanks for documenting all of the issues, this is helpful.

And I agree with your assessment that this seems more like people jumping into flat config without doing any bit of reading beforehand. I see these as primarily learn-once problems, meaning that once people learn how it works, they're unlikely to repeat the mistake. 

It's become very apparent to me throughout this process that most users do zero reading when attempting to convert to flat config. A shockingly common occurrence is people simply renaming `.eslintrc.js` to `eslint.config.js` and then contacting us because "it doesn't work." So at the moment, I'm inclined to tell people to RTFM vs. make any drastic changes.

That said, eslintrc wasn't born in a single release, and I'm sure flat config will continue to evolve as we gather feedback and get more experience with it. Flat config isn't "done", but it is "baked", and I think we need to continue to evaluate it as we go.
